packages/remote-feature-flag-controller/src/remote-feature-flag-controller.ts

```diff
@@ -0,0 +1,215 @@
+ *
+ * @returns The default controller state.
+ */
+export function getDefaultRemoteFeatureFlagControllerState(): RemoteFeatureFlagControllerState {
+  return {
+    remoteFeatureFlag: [],
+    cacheTimestamp: 0,
+  };
+}
+
+/**
+ * The RemoteFeatureFlagController manages the retrieval and caching of remote feature flags.
+ * It fetches feature flags from a remote API, caches them, and provides methods to access
+ * and manage these flags. The controller ensures that feature flags are refreshed based on
+ * a specified interval and handles cases where the controller is disabled or the network is unavailable.
+ */
+export class RemoteFeatureFlagController extends BaseController<
+  typeof controllerName,
+  RemoteFeatureFlagControllerState,
+  RemoteFeatureFlagControllerMessenger
+> {
+  readonly #fetchInterval: number;
+
+  #disabled: boolean;
+
+  #clientConfigApiService: AbstractClientConfigApiService;
+
+  #inProgressFlagUpdate?: Promise<FeatureFlags>;
+
+  /**
+   * Constructs a new RemoteFeatureFlagController instance.
+   *
+   * @param options - The controller options.
+   * @param options.messenger - The controller messenger used for communication.
+   * @param options.state - The initial state of the controller.
+   * @param options.clientConfigApiService - The service instance to fetch remote feature flags.
+   * @param options.fetchInterval - The interval in milliseconds before cached flags expire. Defaults to 1 day.
+   * @param options.disabled - Determines if the controller should be disabled initially. Defaults to false.
+   */
+  constructor({
+    messenger,
+    state,
+    clientConfigApiService,
+    fetchInterval = DEFAULT_CACHE_DURATION,
+    disabled = false,
+  }: {
+    messenger: RemoteFeatureFlagControllerMessenger;
+    state?: Partial<RemoteFeatureFlagControllerState>;
+    clientConfigApiService: AbstractClientConfigApiService;
+    fetchInterval?: number;
+    disabled?: boolean;
+  }) {
+    super({
+      name: controllerName,
+      metadata: remoteFeatureFlagControllerMetadata,
+      messenger,
+      state: {
+        ...getDefaultRemoteFeatureFlagControllerState(),
+        ...state,
+      },
+    });
+
+    this.#fetchInterval = fetchInterval;
+    this.#disabled = disabled;
+    this.#clientConfigApiService = clientConfigApiService;
+  }
+
+  /**
+   * Checks if the cached feature flags are expired based on the fetch interval.
+   *
+   * @returns Whether the cache is expired (`true`) or still valid (`false`).
+   * @private
+   */
+  #isCacheExpired(): boolean {
+    return Date.now() - this.state.cacheTimestamp > this.#fetchInterval;
+  }
+
+  /**
+   * Retrieves the remote feature flags, fetching from the API if necessary.
+   * Uses caching to prevent redundant API calls and handles concurrent fetches.
+   *
+   * @returns A promise that resolves to the current set of feature flags.
+   */
+  async getRemoteFeatureFlag(): Promise<FeatureFlags> {
+    if (this.#disabled) {
+      return [];
```

Comment:
Should we return the cached flags here instead?

Reply:
I agree with returning cached data when disabled is true. It was changed to worked that way in d0b81fafd2b50f50a88ec0d25a42a1ff3aa8d498